### PR TITLE
vLLM always places spaces between special tokens

### LINKED ISSUE #1347: vLLM always places spaces between special tokens
I have a model that in certain cases is expected to generate sequences of special tokens.  When I call in to vLLM the expected special tokens in the `text` of the response have spaces between them.  This is caused by an incomplete (when compared with the adapted huggingface version) API for `_convert_tokens_to_string_with_added_encoders`.  The referenced implementation (see below) allows for the caller to pass a `spaces_between_special_tokens: bool` but this option is absent from the vLLM implementation.

```
    # Adapted from
    # https://github.com/huggingface/transformers/blob/v4.28.0/src/transformers/tokenization_utils.py#L921
    # NOTE(woosuk): The following code is slow because it runs a for loop over
    # the output_tokens. In Python, running a for loop over a list can be slow
    # even when the loop body is very simple.```

### PR DESCRIPTION
**Pull request by blahblahasdf**
Implement huggingface's `spaces_between_special_tokens` in `_convert_tokens_to_string_with_added_encoders`.

Fixes #1347

### CODE REVIEW

**Review by simon-mo — COMMENTED** [review]
Thank you for your contribution, can you undo the naming change `request -> req` so the diff is cleaner to view?

**Comment by blahblahasdf** [comment]
> Thank you for your contribution, can you undo the naming change `request -> req` so the diff is cleaner to view?

Yes, no problem.  Done.

File: vllm/entrypoints/openai/api_server.py [hunk 1]
```diff
@@ -212,6 +212,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
     request_id = f"cmpl-{random_uuid()}"
     created_time = int(time.monotonic())
     try:
+        spaces_between_special_tokens = request.spaces_between_special_tokens
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:215]:
```suggestion
```

File: vllm/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -226,6 +227,7 @@ async def create_chat_completion(request: ChatCompletionRequest,
             ignore_eos=request.ignore_eos,
             use_beam_search=request.use_beam_search,
             skip_special_tokens=request.skip_special_tokens,
+            spaces_between_special_tokens=spaces_between_special_tokens,
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:230]:
```suggestion
            spaces_between_special_tokens=request.spaces_between_special_tokens,
```

File: vllm/entrypoints/openai/api_server.py [hunk 3]
```diff
@@ -413,6 +415,7 @@ async def create_completion(request: CompletionRequest, raw_request: Request):
 
     created_time = int(time.monotonic())
     try:
+        spaces_between_special_tokens = request.spaces_between_special_tokens
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:418]:
```suggestion
```

File: vllm/entrypoints/openai/api_server.py [hunk 4]
```diff
@@ -428,6 +431,7 @@ async def create_completion(request: CompletionRequest, raw_request: Request):
             logprobs=request.logprobs,
             use_beam_search=request.use_beam_search,
             skip_special_tokens=request.skip_special_tokens,
+            spaces_between_special_tokens=spaces_between_special_tokens,
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:434]:
```suggestion
            spaces_between_special_tokens=request.spaces_between_special_tokens,
```

**Review by simon-mo — APPROVED** [review]

File: vllm/entrypoints/openai/api_server.py (hunk 2 above)
Comment (blahblahasdf) [reply vllm/entrypoints/openai/api_server.py:230]: I would 100% love to take this suggestion but that's 81 characters long which violates the codestyle and causes the build to break.

**Pull request merged**